common/djangoapps/student/views.py

```diff
@@ -357,24 +357,29 @@ def signin_user(request):
     external_auth_response = external_auth_login(request)
     if external_auth_response is not None:
         return external_auth_response
+    # Determine the URL to redirect to following login:
+    redirect_to = get_next_url_for_login_page(request)
     if request.user.is_authenticated():
-        return redirect(reverse('dashboard'))
+        return redirect(redirect_to)
+
+    third_party_auth_error = None
+    for msg in messages.get_messages(request):
+        if msg.extra_tags.split()[0] == "social-auth":
+            third_party_auth_error = unicode(msg)
```

Comment:
@bradenmacdonald, I agree with your approach to this. However, I want to make sure @sarina is aware that this has the potential to allow untranslated strings into the system, and that she's on board with it. @sarina?